nucleus/core/bootstrap/src/main/java/com/sun/enterprise/glassfish/bootstrap/GlassFishImpl.java

```diff
@@ -30,31 +32,41 @@
 /**
  * @author Sanjeeb.Sahoo@Sun.COM
  */
-
 public class GlassFishImpl implements GlassFish {
 
+    private static final String CONFIG_PROP_PREFIX = "embedded-glassfish-config.";
+
     private ModuleStartup gfKernel;
-    private ServiceLocator habitat;
-    volatile Status status = Status.INIT;
+    private ServiceLocator serviceLocator;
+    private volatile Status status = Status.INIT;
```

Comment:
Please move this initialization into ctor after `serviceLocator` assignment. This will guarantee thread visibility of the `serviceLocator` and `gfKernel` values.

Reply:
I moved that to the constructor, but as the first line. It doesn't matter, if constructor fails, its fields will not be used.

Reply:
In this case `start()` method may throw NPE if called from another thread. Field initialization done after `super()` but before fields assignment in constructor. We have next sequence: init `gfKernel` with default value (null) -> volatile write `state`-> `gfKernel` assignment -> volatile read `state` in `start()` method -> volatile write `state` -> plain read `gfKernel` (at the line 75). This assignment does not cover by the happens-before relation. And another thread can see null. Synchronized on `start` does not matters because `gfKernel` initialized without synchronization.

Reply:
I am not sure if we understand each other, so maybe silly question - does it apply for the current code, when I moved it to the constructor? I believe then all three fields should be initialized in order.
I understand and absolutely agree with you about initialization outside constructor. Btw on other my projects tools prohibit calling nonstatic methods of the same instance from constructor - for similar reasons.

Reply:
Is it possible to use an object with unfinished constructor in another thread? How could that happen? The only such scenario I can imagine is if we would start thread using `this` in constructor.
I really want to get it.

Reply:
Aha, seems I forgot something important again ... you are right, thanks!
https://stackoverflow.com/a/10528614/375449